vivarium/simulator/states.py

```diff
@@ -0,0 +1,316 @@
+    return jnp.array(list(mcolors.to_rgb(color_str)))
+
+
+def init_simulator_state(
+        box_size: float = 100.,
+        n_agents: int = 10,
+        n_objects: int = 2,
+        num_steps_lax: int = 4,
+        dt: float = 0.1,
+        freq: float = 40.,
+        neighbor_radius: float = 100.,
+        to_jit: bool = True,
+        use_fori_loop: bool = False
+        ) -> SimulatorState:
+    """
+    Initialize simulator state with given parameters
+    """
+    return SimulatorState(
+        idx=jnp.array([0]),
+        box_size=jnp.array([box_size]),                       
+        n_agents=jnp.array([n_agents]),
+        n_objects=jnp.array([n_objects]),
+        num_steps_lax=jnp.array([num_steps_lax], dtype=int),
+        dt=jnp.array([dt], dtype=float),
+        freq=jnp.array([freq], dtype=float),
+        neighbor_radius=jnp.array([neighbor_radius], dtype=float),
+        # Use 1*bool to transform True to 1 and False to 0
+        to_jit= jnp.array([1*to_jit]),
+        use_fori_loop=jnp.array([1*use_fori_loop]))
+
+
+def _init_positions(key_pos, positions, n_elements, box_size, n_dims=2):
+    assert (len(positions) == n_elements if positions else True)
+    # If positions are passed, transform them in jax array
+    if positions:
+        positions = jnp.array(positions)
+    # Else initialize random positions
+    else:
+        positions = random.uniform(key_pos, (n_elements, n_dims)) * box_size
+    return positions
+
+def _init_existing(n_existing, n_elements):
+    if n_existing:
+        assert n_existing <= n_elements
+        existing_arr = jnp.ones((n_existing))
+        non_existing_arr = jnp.zeros((n_elements - n_existing))
+        exists_array = jnp.concatenate((existing_arr, non_existing_arr))
+    else:
+        exists_array = jnp.ones((n_elements))
+    return exists_array
+
+# TODO : Should also add union float, list[float] for friction, diameter ..
+# Also think it would be easier to only handle the nve state in the physics engine. (if it doesn't make the  simulation run slower)
+# Here it makes a really long function, and it isn't very modular 
+# Plus we store different attributes of entities (e.g Agents positions and sensors) in different dataclasses
+def init_nve_state(
+        simulator_state: SimulatorState,
+        diameter: float = 5.,
+        friction: float = 0.1,
+        mass_center: float = 1.,
+        mass_orientation: float = 0.125,
+        agents_positions: Optional[Union[List[float], None]] = None,
+        objects_positions: Optional[Union[List[float], None]] = None,
+        existing_agents: Optional[Union[int, List[float], None]] = None,
+        existing_objects: Optional[Union[int, List[float], None]] = None,
+        seed: int = 0,
+        ) -> NVEState:
+    """
+    Initialize nve state with given parameters
+    """
+    n_agents = simulator_state.n_agents[0]
+    n_objects = simulator_state.n_objects[0]
+    n_entities = n_agents + n_objects
+
+    key = random.PRNGKey(seed)
```

Comment:
I'm not sure this follows the good practice about how to deal with randomness in JAX, see https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html

Typically, the seed should be defined in a single place in the code, and all functions requiring a random number should use `jax.random.split`. `PRNGKey` should be called only once in the entire code base. (it's how I understand it at least) 

See if you can fix it here, otherwise create an issue (point at the code here so that we don't forget it)

Reply:
It is a good question. I am not sure about this (because in our case this key only serves to initialize random positions, and they key in simulator serve other purposes) but it will be easy to change if we want to. Actually I checked some RL repos in Jax and they seem to define keys with seeds in different parts of the code (ex: in environments and in algorithms). 

From what I saw we only have another key initialization in state, we can just change the init methods so it doesn't accept a seed that generates a key, but directly a key (that you generate at the  beginning of the script you run).